core_types/src/dna/fn_declarations.rs

```diff
@@ -0,0 +1,115 @@
+//! File holding all the structs for handling function declarations defined in DNA.
+
+use crate::dna::capabilities::CapabilityType;
+
+/// Represents the type declaration for zome function parameter
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, Hash)]
+pub struct FnParameter {
+    #[serde(rename = "type")]
+    pub parameter_type: String,
+    pub name: String,
+}
+
+impl FnParameter {
+    #[allow(dead_code)]
+    pub fn new<S: Into<String>>(n: S, t: S) -> FnParameter {
+        FnParameter {
+            name: n.into(),
+            parameter_type: t.into(),
+        }
+    }
+}
+
+/// Represents a zome function declaration
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, Hash)]
+pub struct FnDeclaration {
+    /// The name of this fn declaration.
+    #[serde(default)]
+    pub name: String,
+    pub inputs: Vec<FnParameter>,
+    pub outputs: Vec<FnParameter>,
+}
+
+impl Default for FnDeclaration {
+    /// Defaults for a "fn_declarations" object.
+    fn default() -> Self {
+        FnDeclaration {
+            name: String::new(),
+            inputs: Vec::new(),
+            outputs: Vec::new(),
+        }
+    }
+}
+
+impl FnDeclaration {
+    /// Allow sane defaults for `FnDecrlaration::new()`.
+    pub fn new() -> Self {
+        Default::default()
+    }
+}
+
+/// Represents an trait definition for bridging
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, Hash)]
+pub struct Trait {
+    /// capability type enum
+    #[serde(rename = "type")]
+    pub cap_type: CapabilityType,
+
+    /// "functions" array
+    #[serde(default)]
+    pub functions: Vec<FnDeclaration>,
```

Comment:
should this be called `fn_declarations` also?

Reply:
So the reason I didn't call it declarations here is that it isn't where the functions are declared, its where the trait is saying that "I need a function that looks like this".  There's only one place where it's declared, and that's in the zome.  Possible we could make a distinctions between declared and defined...  I think I'm pushing this to our DND session.